src/interactivity/base/HostSignalInputThread.cpp

```diff
@@ -0,0 +1,240 @@
+{
+    HostSignals signalId;
+    while (_GetData(&signalId, sizeof(signalId)))
+    {
+        switch (signalId)
+        {
+        case HostSignals::NotifyApp:
+        {
+            auto msg = _ReceiveTypedPacket<HostSignalNotifyAppData>();
+
+            LOG_IF_NTSTATUS_FAILED(ServiceLocator::LocateConsoleControl()->NotifyConsoleApplication(msg.processId));
+
+            break;
+        }
+        case HostSignals::SetForeground:
+        {
+            auto msg = _ReceiveTypedPacket<HostSignalSetForegroundData>();
+
+            LOG_IF_NTSTATUS_FAILED(ServiceLocator::LocateConsoleControl()->SetForeground(ULongToHandle(msg.processId), msg.isForeground));
+
+            break;
+        }
+        case HostSignals::EndTask:
+        {
+            auto msg = _ReceiveTypedPacket<HostSignalEndTaskData>();
+
+            LOG_IF_NTSTATUS_FAILED(ServiceLocator::LocateConsoleControl()->EndTask(ULongToHandle(msg.processId), msg.eventType, msg.ctrlFlags));
+
+            break;
+        }
+        default:
+        {
+            THROW_HR(E_UNEXPECTED);
+            break;
+        }
+        }
+    }
+    return S_OK;
+}
+
+// Method Description:
+// - Skips the file stream forward by the specified number of bytes.
+// Arguments:
+// - cbBytes - Count of bytes to skip forward
+// Return Value:
+// - True if we could skip forward successfully. False otherwise.
+bool HostSignalInputThread::_AdvanceReader(const DWORD cbBytes)
+{
+    std::array<BYTE, 8> buffer;
+    auto bytesRemaining = cbBytes;
+
+    while (bytesRemaining > 0)
+    {
+        const auto advance = std::min(bytesRemaining, gsl::narrow_cast<DWORD>(buffer.max_size()));
+
+        if (!_GetData(buffer.data(), advance))
+        {
+            return false;
+        }
+
+        bytesRemaining -= advance;
+    }
+
+    return true;
+}
+
+// Method Description:
+// - Retrieves bytes from the file stream and exits or throws errors should the pipe state
+//   be compromised.
+// Arguments:
+// - pBuffer - Buffer to fill with data.
+// - cbBuffer - Count of bytes in the given buffer.
+// Return Value:
+// - True if data was retrieved successfully. False otherwise.
+bool HostSignalInputThread::_GetData(_Out_writes_bytes_(cbBuffer) void* const pBuffer,
+                                     const DWORD cbBuffer)
+{
+    DWORD dwRead = 0;
+    // If we failed to read because the terminal broke our pipe (usually due
+    //      to dying itself), close gracefully with ERROR_BROKEN_PIPE.
+    // Otherwise throw an exception. ERROR_BROKEN_PIPE is the only case that
+    //       we want to gracefully close in.
+    if (FALSE == ReadFile(_hFile.get(), pBuffer, cbBuffer, &dwRead, nullptr))
+    {
+        DWORD lastError = GetLastError();
+        if (lastError == ERROR_BROKEN_PIPE)
+        {
+            _Shutdown();
+            return false;
+        }
+        else
+        {
+            THROW_WIN32(lastError);
+        }
+    }
+    else if (dwRead != cbBuffer)

```

Comment:
You don't need the `else`. This condition isn't related to the "ReadFile failed"-handling code anymore anyways. (You could also remove the else above.)